File: cpp/chronomesh/src/workflow.cpp
#include "chronomesh/core.hpp"
#include <deque>

namespace chronomesh {

// ---------------------------------------------------------------------------
// State transition graph
// ---------------------------------------------------------------------------


static const std::map<std::string, std::map<std::string, bool>> graph = {
    {"queued",    {{"allocated", true}, {"cancelled", true}}},
    {"allocated", {{"departed", true}, {"cancelled", true}}},
    {"departed",  {{"arrived", true}, {"cancelled", true}}},
    {"arrived",   {}},
};


static const std::map<std::string, bool> terminal_states = {
    {"cancelled", true},
};

// ---------------------------------------------------------------------------
// Core transition validation
// ---------------------------------------------------------------------------

bool can_transition(const std::string& from, const std::string& to) {
  auto it = graph.find(from);
  if (it == graph.end()) return false;
  auto jt = it->second.find(to);
  return jt != it->second.end() && jt->second;
}

// ---------------------------------------------------------------------------
// Transition helpers
// ---------------------------------------------------------------------------

std::vector<std::string> allowed_transitions(const std::string& from) {
  auto it = graph.find(from);
  if (it == graph.end()) return {};
  std::vector<std::string> result;
  for (const auto& [state, _] : it->second) result.push_back(state);
  return result;
}

bool is_valid_state(const std::string& state) {
  return graph.count(state) > 0 || terminal_states.count(state) > 0;
}

bool is_terminal_state(const std::string& state) {
  return terminal_states.count(state) > 0;
}

// ---------------------------------------------------------------------------
// Shortest path (BFS)
// ---------------------------------------------------------------------------

std::vector<std::string> shortest_path(const std::string& from, const std::string& to) {
  if (from == to) return {from};
  std::map<std::string, bool> visited;
  visited[from] = true;
  std::deque<std::vector<std::string>> queue;
  queue.push_back({from});

  while (!queue.empty()) {
    auto path = queue.front();
    queue.pop_front();
    auto current = path.back();
    auto it = graph.find(current);
    if (it == graph.end()) continue;
    for (const auto& [next, _] : it->second) {
      if (next == to) {
        path.push_back(next);
        return path;
      }
      if (!visited[next]) {
        visited[next] = true;
        auto new_path = path;
        new_path.push_back(next);
        queue.push_back(new_path);
      }
    }
  }
  return {};
}

// ---------------------------------------------------------------------------
// Workflow engine — manages entity lifecycles
// ---------------------------------------------------------------------------

WorkflowEngine::WorkflowEngine() {}

bool WorkflowEngine::register_entity(const std::string& entity_id, const std::string& initial_state) {
  std::lock_guard lock(mu_);
  std::string state = initial_state.empty() ? "queued" : initial_state;
  if (graph.find(state) == graph.end()) return false;
  entities_[entity_id] = Entity{state, {}};
  return true;
}

std::string WorkflowEngine::get_state(const std::string& entity_id) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) return "";
  return it->second.state;
}

TransitionResult WorkflowEngine::transition(const std::string& entity_id, const std::string& to) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) {
    return TransitionResult{false, "entity_not_found", "", to};
  }
  auto& entity = it->second;
  if (!can_transition(entity.state, to)) {
    return TransitionResult{false, "invalid_transition", entity.state, to};
  }
  TransitionRecord record{entity_id, entity.state, to};
  entity.transitions.push_back(record);
  entity.state = to;
  log_.push_back(record);
  return TransitionResult{true, "", record.from, to};
}

bool WorkflowEngine::is_terminal(const std::string& entity_id) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) return false;
  return terminal_states.count(it->second.state) > 0;
}

int WorkflowEngine::active_count() {
  std::lock_guard lock(mu_);
  int count = 0;
  for (const auto& [_, entity] : entities_) {
    if (terminal_states.count(entity.state) == 0) count++;
  }
  return count;
}

std::vector<TransitionRecord> WorkflowEngine::entity_history(const std::string& entity_id) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) return {};
  return it->second.transitions;
}

std::vector<TransitionRecord> WorkflowEngine::audit_log() {
  std::lock_guard lock(mu_);
  return log_;
}

// ---------------------------------------------------------------------------
// Force-complete an entity to arrived state
// ---------------------------------------------------------------------------

bool WorkflowEngine::force_complete(const std::string& entity_id) {
  std::lock_guard lock(mu_);
  auto it = entities_.find(entity_id);
  if (it == entities_.end()) return false;
  auto& entity = it->second;
  if (terminal_states.count(entity.state) > 0) return false;

  auto path = shortest_path(entity.state, "arrived");
  if (path.empty()) return false;

  TransitionRecord record{entity_id, path.front(), path.back()};
  entity.transitions.push_back(record);
  log_.push_back(record);
  entity.state = "arrived";
  return true;
}

// ---------------------------------------------------------------------------
// Bulk transition for multiple entities
// ---------------------------------------------------------------------------

std::vector<TransitionResult> WorkflowEngine::bulk_transition(
    const std::vector<std::string>& entity_ids, const std::string& to) {
  std::lock_guard lock(mu_);
  std::vector<TransitionResult> results;

  for (const auto& eid : entity_ids) {
    auto it = entities_.find(eid);
    if (it == entities_.end()) {
      results.push_back(TransitionResult{false, "entity_not_found", "", to});
      continue;
    }
    auto& entity = it->second;
    if (!can_transition(entity.state, to)) {
      results.push_back(TransitionResult{false, "invalid_transition", entity.state, to});
      continue;
    }
    std::string prev = entity.state;
    TransitionRecord rec{eid, prev, to};
    entity.transitions.push_back(rec);
    entity.state = to;
    log_.push_back(rec);
    results.push_back(TransitionResult{true, "", prev, to});
  }
  return results;
}

// ---------------------------------------------------------------------------
// Count terminal entities
// ---------------------------------------------------------------------------

int WorkflowEngine::terminal_count() {
  std::lock_guard lock(mu_);
  int count = 0;
  for (const auto& [_, entity] : entities_) {
    if (terminal_states.count(entity.state) > 0) count++;
  }
  return count;
}

}

File: go/atlasdispatch/internal/workflow/workflow.go
package workflow

import (
	"sync"
	"time"
)

// ---------------------------------------------------------------------------
// State transition graph
// ---------------------------------------------------------------------------

var graph = map[string]map[string]bool{
	"queued":    {"allocated": true, "cancelled": true},
	"allocated": {"departed": true, "cancelled": true},
	"departed":  {"arrived": true},
	"arrived":   {},
}

var terminalStates = map[string]bool{
	"arrived":   true,
	"cancelled": true,
}

// ---------------------------------------------------------------------------
// Core transition validation
// ---------------------------------------------------------------------------

func CanTransition(from, to string) bool {
	
	return graph[from][to]
}

// ---------------------------------------------------------------------------
// Transition helpers
// ---------------------------------------------------------------------------

func AllowedTransitions(from string) []string {
	allowed := graph[from]
	result := make([]string, 0, len(allowed))
	for state := range allowed {
		result = append(result, state)
	}
	return result
}

func IsValidState(state string) bool {
	
	_, ok := graph[state]
	return ok || terminalStates[state]
}

func IsTerminalState(state string) bool {
	return terminalStates[state]
}

// ---------------------------------------------------------------------------
// Shortest path (BFS)
// ---------------------------------------------------------------------------

func ShortestPath(from, to string) []string {
	
	if from == to {
		return []string{from}
	}
	visited := map[string]bool{from: true}
	queue := [][]string{{from}}
	for len(queue) > 0 {
		path := queue[0]
		queue = queue[1:]
		current := path[len(path)-1]
		for next := range graph[current] {
			if next == to {
				return append(path, next)
			}
			if !visited[next] {
				visited[next] = true
				newPath := make([]string, len(path)+1)
				copy(newPath, path)
				newPath[len(path)] = next
				queue = append(queue, newPath)
			}
		}
	}
	return nil
}

// ---------------------------------------------------------------------------
// Workflow engine — manages entity lifecycles
// ---------------------------------------------------------------------------

type TransitionRecord struct {
	EntityID string
	From     string
	To       string
	At       time.Time
}

type entity struct {
	state       string
	transitions []TransitionRecord
}

type WorkflowEngine struct {
	mu       sync.Mutex
	entities map[string]*entity
	log      []TransitionRecord
}

func NewWorkflowEngine() *WorkflowEngine {
	return &WorkflowEngine{
		entities: make(map[string]*entity),
	}
}

func (we *WorkflowEngine) Register(entityID, initialState string) error {
	we.mu.Lock()
	defer we.mu.Unlock()
	
	if initialState == "" {
		initialState = "allocated"
	}
	if _, ok := graph[initialState]; !ok {
		return &InvalidStateError{State: initialState}
	}
	we.entities[entityID] = &entity{state: initialState}
	return nil
}

func (we *WorkflowEngine) GetState(entityID string) string {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return ""
	}
	return e.state
}

type TransitionResult struct {
	Success bool
	Reason  string
	From    string
	To      string
}

func (we *WorkflowEngine) Transition(entityID, to string) TransitionResult {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return TransitionResult{Success: false, Reason: "entity_not_found"}
	}
	if !CanTransition(e.state, to) {
		return TransitionResult{Success: false, Reason: "invalid_transition", From: e.state, To: to}
	}
	record := TransitionRecord{
		EntityID: entityID,
		From:     e.state,
		To:       to,
		At:       time.Now(),
	}
	e.transitions = append(e.transitions, record)
	e.state = to
	we.log = append(we.log, record)
	return TransitionResult{Success: true, From: record.From, To: to}
}

func (we *WorkflowEngine) IsTerminal(entityID string) bool {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return false
	}
	return terminalStates[e.state]
}

func (we *WorkflowEngine) ActiveCount() int {
	we.mu.Lock()
	defer we.mu.Unlock()
	count := 0
	for _, e := range we.entities {
		if !terminalStates[e.state] {
			count++
		}
	}
	return count
}

func (we *WorkflowEngine) History(entityID string) []TransitionRecord {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return nil
	}
	result := make([]TransitionRecord, len(e.transitions))
	copy(result, e.transitions)
	return result
}

func (we *WorkflowEngine) AuditLog() []TransitionRecord {
	we.mu.Lock()
	defer we.mu.Unlock()
	result := make([]TransitionRecord, len(we.log))
	copy(result, we.log)
	return result
}

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

type InvalidStateError struct {
	State string
}

func (e *InvalidStateError) Error() string {
	return "invalid state: " + e.State
}

// ---------------------------------------------------------------------------
// Path validation — checks if a sequence of states forms a valid path
// ---------------------------------------------------------------------------

func ValidateTransitionPath(path []string) bool {
	if len(path) < 2 {
		return len(path) == 1 && IsValidState(path[0])
	}
	for i := 0; i < len(path)-1; i += 2 {
		if !CanTransition(path[i], path[i+1]) {
			return false
		}
	}
	return true
}

// ---------------------------------------------------------------------------
// Batch transitions — transition multiple entities atomically
// ---------------------------------------------------------------------------

func (we *WorkflowEngine) TransitionBatch(transitions map[string]string) map[string]TransitionResult {
	we.mu.Lock()
	defer we.mu.Unlock()
	results := make(map[string]TransitionResult)
	for entityID, to := range transitions {
		e, ok := we.entities[entityID]
		if !ok {
			results[entityID] = TransitionResult{Success: false, Reason: "entity_not_found"}
			continue
		}
		from := e.state
		if !CanTransition(from, to) {
			results[entityID] = TransitionResult{Success: false, Reason: "invalid_transition", From: from, To: to}
			continue
		}
		record := TransitionRecord{
			EntityID: entityID,
			From:     from,
			To:       to,
			At:       time.Now(),
		}
		e.state = to
		e.transitions = append(e.transitions, record)
		results[entityID] = TransitionResult{Success: true, From: from, To: to}
	}
	return results
}

// ---------------------------------------------------------------------------
// Atomic transition chain — applies a sequence of transitions to one entity.
// If any transition fails, all previous transitions in the chain are
// rolled back to maintain atomicity.
// ---------------------------------------------------------------------------

func (we *WorkflowEngine) TransitionChain(entityID string, targets []string) []TransitionResult {
	we.mu.Lock()
	defer we.mu.Unlock()
	e, ok := we.entities[entityID]
	if !ok {
		return []TransitionResult{{Success: false, Reason: "entity_not_found"}}
	}
	results := make([]TransitionResult, 0, len(targets))
	for _, target := range targets {
		if !CanTransition(e.state, target) {
			results = append(results, TransitionResult{
				Success: false, Reason: "invalid_transition",
				From: e.state, To: target,
			})
			continue
		}
		record := TransitionRecord{
			EntityID: entityID, From: e.state, To: target, At: time.Now(),
		}
		e.state = target
		e.transitions = append(e.transitions, record)
		we.log = append(we.log, record)
		results = append(results, TransitionResult{Success: true, From: record.From, To: target})
	}
	return results
}
